File: DAI/BasicDAIPlayer-iOS/objc/BasicDAIPlayer/BaseViewController.m
//
//  BaseViewController.m
//  BasicDAIPlayer
//
//  Copyright © 2023 Brightcove, Inc. All rights reserved.
//

#import <AppTrackingTransparency/AppTrackingTransparency.h>

#import <BrightcovePlayerSDK/BrightcovePlayerSDK.h>
#import <BrightcoveDAI/BrightcoveDAI.h>

#import <GoogleInteractiveMediaAds/GoogleInteractiveMediaAds.h>

#import "BaseViewController.h"


// ** Customize these values with your own account information **
static NSString * const kViewControllerPlaybackServicePolicyKey = @"BCpkADawqM0T8lW3nMChuAbrcunBBHmh4YkNl5e6ZrKQwPiK_Y83RAOF4DP5tyBF_ONBVgrEjqW6fbV0nKRuHvjRU3E8jdT9WMTOXfJODoPML6NUDCYTwTHxtNlr5YdyGYaCPLhMUZ3Xu61L";
static NSString * const kViewControllerAccountID = @"5434391461001";
static NSString * const kViewControllerVideoID = @"1753980443013591663";

NSString * const kViewControllerGoogleDAISourceId = @"2528370";
NSString * const kViewControllerGoogleDAIVideoId = @"tears-of-steel";
NSString * const kViewControllerGoogleDAIAssetKey = @"sN_IYUG8STe1ZzhIIE_ksA";


@implementation BaseViewController

- (void)dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:self.notificationReceipt];
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    if (@available(iOS 14, *))
    {
        [ATTrackingManager requestTrackingAuthorizationWithCompletionHandler:^(ATTrackingManagerAuthorizationStatus status) {
            dispatch_async(dispatch_get_main_queue(), ^{
                [self setup];
            });
        }];
    }
    else
    {
        [self setup];
    }
}

- (void)setup
{
    self.manager = BCOVPlayerSDKManager.sharedManager;

    BCOVPlaybackService *playbackService = ({
        BCOVPlaybackServiceRequestFactory *factory = [[BCOVPlaybackServiceRequestFactory alloc] initWithAccountId:kViewControllerAccountID
                                                                                                        policyKey:kViewControllerPlaybackServicePolicyKey];
        BCOVPlaybackService *service = [[BCOVPlaybackService alloc] initWithRequestFactory:factory];
        service;
    });

    self.playbackService = playbackService;

    BCOVPUIPlayerView *playerView = ({
        BCOVPUIPlayerViewOptions *options = [[BCOVPUIPlayerViewOptions alloc] init];
        options.presentingViewController = self;
        options.automaticControlTypeSelection = YES;

        BCOVPUIPlayerView *playerView = [[BCOVPUIPlayerView alloc] initWithPlaybackController:nil
                                                                                      options:options
                                                                                 controlsView:nil];
        playerView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
        playerView.frame = self.videoContainerView.bounds;
        [self.videoContainerView addSubview:playerView];
        playerView;
    });

    self.playerView = playerView;

    [self setupPlaybackController];

    [self resumeAdAfterForeground];

    [self requestContentFromPlaybackService];
}

- (void)setupPlaybackController
{
    // NO-OP
}

- (BCOVVideo *)updateVideo:(BCOVVideo *)video
{
    return video;
}

- (void)resumeAdAfterForeground
{
    // When the app goes to the background, the Google IMA library will pause
    // the ad. This code demonstrates how you would resume the ad when entering
    // the foreground.

    __weak typeof(self) weakSelf = self;

    self.notificationReceipt = [[NSNotificationCenter defaultCenter] addObserverForName:UIApplicationWillEnterForegroundNotification
                                                                                 object:nil
                                                                                  queue:nil
                                                                             usingBlock:^(NSNotification *note) {
        
        __strong typeof(weakSelf) strongSelf = weakSelf;
        
        if (strongSelf.adIsPlaying)
        {
            [strongSelf.playbackController resumeAd];
        }

    }];
}

- (void)requestContentFromPlaybackService
{
    NSDictionary *configuration = @{
        kBCOVPlaybackServiceConfigurationKeyAssetID: kViewControllerVideoID
    };

    __weak typeof(self) weakSelf = self;

    [self.playbackService findVideoWithConfiguration:configuration
                                     queryParameters:nil
                                          completion:^(BCOVVideo *video, NSDictionary *jsonResponse, NSError *error) {

        __strong typeof(weakSelf) strongSelf = weakSelf;

        if (video)
        {
            BCOVVideo *updatedVideo = [strongSelf updateVideo:video];

            [strongSelf.playbackController setVideos:@[ updatedVideo ]];
        }
        else
        {
            NSLog(@"ViewController Debug - Error retrieving video");
        }
    }];
}


#pragma mark BCOVPlaybackControllerDelegate

- (void)playbackController:(id<BCOVPlaybackController>)controller didAdvanceToPlaybackSession:(id<BCOVPlaybackSession>)session
{
    NSLog(@"ViewController - Advanced to new session.");

    // Enable route detection for AirPlay
    // https://developer.apple.com/documentation/avfoundation/avroutedetector/2915762-routedetectionenabled
    self.playerView.controlsView.routeDetector.routeDetectionEnabled = YES;
}

- (void)playbackController:(id<BCOVPlaybackController>)controller playbackSession:(id<BCOVPlaybackSession>)session didReceiveLifecycleEvent:(BCOVPlaybackSessionLifecycleEvent *)lifecycleEvent
{
    // Ad events are emitted by the BCOVIMA plugin through lifecycle events.
    // The events are defined BCOVDAIComponent.h.
    NSString *type = lifecycleEvent.eventType;

    if ([type isEqualToString:kBCOVDAILifecycleEventAdsLoaderLoaded])
    {
        NSLog(@"ViewController - Ads loaded.");
    }
    else if ([type isEqualToString:kBCOVDAILifecycleEventAdsManagerDidReceiveAdEvent])
    {
        IMAAdEvent *adEvent = lifecycleEvent.properties[@"adEvent"];
        switch (adEvent.type)
        {
            case kIMAAdEvent_STARTED:
                NSLog(@"ViewController - Ad Started.");
                self.adIsPlaying = YES;
                break;
            case kIMAAdEvent_COMPLETE:
                NSLog(@"ViewController - Ad Completed.");
                self.adIsPlaying = NO;
                break;
            case kIMAAdEvent_ALL_ADS_COMPLETED:
                NSLog(@"ViewController - All ads completed.");
                break;
            default:
                break;
        }
    }
    else if ([type isEqualToString:kBCOVPlaybackSessionLifecycleEventEnd])
    {
        // Disable route detection for AirPlay
        // https://developer.apple.com/documentation/avfoundation/avroutedetector/2915762-routedetectionenabled
        self.playerView.controlsView.routeDetector.routeDetectionEnabled = NO;
    }
}


#pragma mark IMALinkOpenerDelegate

- (void)linkOpenerWillOpenInAppLink:(NSObject *)linkOpener
{
    NSLog(@"IMALinkOpenerDelegate: In-app browser will open");
}

- (void)linkOpenerDidOpenInAppLink:(NSObject *)linkOpener
{
    NSLog(@"IMALinkOpenerDelegate: In-app browser did open");
}

- (void)linkOpenerWillCloseInAppLink:(NSObject *)linkOpener
{
    NSLog(@"IMALinkOpenerDelegate: In-app browser will close");
}

- (void)linkOpenerDidCloseInAppLink:(NSObject *)linkOpener
{
    NSLog(@"IMALinkOpenerDelegate: In-app browser did close");
    
    if (self.adIsPlaying)
    {
        [self.playbackController resumeAd];
    }
}

@end

File: DAI/BasicDAIPlayer-iOS/BasicDAIPlayer/BaseViewController.swift
//
//  BaseViewController.swift
//  BasicDAIPlayer
//
//  Copyright © 2026 Brightcove, Inc. All rights reserved.
//

import AdSupport
import AppTrackingTransparency
import UIKit
import GoogleInteractiveMediaAds
import BrightcoveDAI


// Customize these values with your own account information
// Add your Brightcove account and video information here.
let kAccountId = "5434391461001"
let kPolicyKey = "BCpkADawqM0T8lW3nMChuAbrcunBBHmh4YkNl5e6ZrKQwPiK_Y83RAOF4DP5tyBF_ONBVgrEjqW6fbV0nKRuHvjRU3E8jdT9WMTOXfJODoPML6NUDCYTwTHxtNlr5YdyGYaCPLhMUZ3Xu61L"
let kVideoId = "1753980443013591663"


let kGoogleDAISourceId = "2528370"
let kGoogleDAIVideoId = "tears-of-steel"
let kGoogleDAIAssetKey = "sN_IYUG8STe1ZzhIIE_ksA"



class BaseViewController: UIViewController {

    @IBOutlet fileprivate weak var videoContainerView: UIView!

    fileprivate lazy var playbackService: BCOVPlaybackService = {
        let factory = BCOVPlaybackServiceRequestFactory(withAccountId: kAccountId,
                                                        policyKey: kPolicyKey)
        return .init(withRequestFactory: factory)
    }()

    fileprivate(set) lazy var playerView: BCOVPUIPlayerView? = {
        let options = BCOVPUIPlayerViewOptions()
        options.presentingViewController = self
        options.automaticControlTypeSelection = true

        guard let playerView = BCOVPUIPlayerView(playbackController: nil,
                                                 options: options,
                                                 controlsView: nil) else {
            return nil
        }

        playerView.delegate = self

        playerView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        playerView.frame = videoContainerView.bounds
        videoContainerView.addSubview(playerView)

        return playerView
    }()

    fileprivate(set) lazy var fps: BCOVPlaybackSessionProvider? = {
        let sdkManager = BCOVPlayerSDKManager.sharedManager()
        let authProxy = BCOVFPSBrightcoveAuthProxy(withPublisherId: nil,
                                                         applicationId: nil)

        return sdkManager.createFairPlaySessionProvider(withAuthorizationProxy: authProxy,
                                                        upstreamSessionProvider: nil)
    }()

    var playbackController: BCOVPlaybackController?

    fileprivate var notificationReceipt: AnyObject?
    fileprivate var adIsPlaying = false

    fileprivate var statusBarHidden = false {
        didSet {
            if let navigationController {
                navigationController.isNavigationBarHidden = statusBarHidden
            }

            navigationItem.hidesBackButton = statusBarHidden

            setNeedsStatusBarAppearanceUpdate()
        }
    }

    override var prefersStatusBarHidden: Bool {
        statusBarHidden
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        setupPlaybackController()

        resumeAdAfterForeground()

        if #available(iOS 14.5, *) {
            ATTrackingManager.requestTrackingAuthorization { _ in
                print("IDFA: \(ASIdentifierManager.shared().advertisingIdentifier.uuidString)")

                DispatchQueue.main.async { [self] in
                    // Tracking authorization completed.
                    // Start loading ads here.
                    requestContentFromPlaybackService()
                }
            }
        } else {
            requestContentFromPlaybackService()
        }
    }

    fileprivate func resumeAdAfterForeground() {

        // When the app goes to the background, the Google IMA library will
        // pause the ad.
        // This code demonstrates how you would resume the ad when entering
        // the foreground.
        notificationReceipt = NotificationCenter.default.addObserver(forName: UIApplication.willEnterForegroundNotification,
                                                                     object: nil,
                                                                     queue: nil) { [weak self] _ in
            guard let self,
                  let playbackController else {
                return
            }

            if adIsPlaying {
                playbackController.resumeAd()
            }
        }
    }

    fileprivate func requestContentFromPlaybackService() {
        let configuration = [BCOVPlaybackService.ConfigurationKeyAssetID: kVideoId]
        playbackService.findVideo(withConfiguration: configuration,
                                  queryParameters: nil) {
            [self] (video: BCOVVideo?,
                    jsonResponse: Any?,
                    error: Error?) in
            guard let video,
                  let playbackController else {
                if let error {
                    print("BaseViewController - Error retrieving video: \(error.localizedDescription)")
                }

                return
            }

#if targetEnvironment(simulator)
            if video.usesFairPlay {
                // FairPlay doesn't work when we're running in a simulator,
                // so put up an alert.
                let alert = UIAlertController(title: "FairPlay Warning",
                                              message: """
                                               FairPlay only works on actual \
                                               iOS or tvOS devices.\n
                                               You will not be able to view \
                                               any FairPlay content in the \
                                               iOS or tvOS simulator.
                                               """,
                                              preferredStyle: .alert)

                alert.addAction(.init(title: "OK", style: .default))

                DispatchQueue.main.async { [self] in
                    present(alert, animated: true)
                }

                return
            }
#endif

            let updatedVideo = updateVideo(video)
            playbackController.setVideos([updatedVideo])
        }
    }

    func setupPlaybackController() {
        // NO-OP
    }

    func updateVideo(_ video: BCOVVideo) -> BCOVVideo {
        return video
    }
}


// MARK: - BCOVPlaybackControllerDelegate

extension BaseViewController: BCOVPlaybackControllerDelegate {

    func playbackController(_ controller: BCOVPlaybackController!,
                            playbackSession session: BCOVPlaybackSession,
                            didReceive lifecycleEvent: BCOVPlaybackSessionLifecycleEvent!) {

        if kBCOVPlaybackSessionLifecycleEventFail == lifecycleEvent.eventType,
           let error = lifecycleEvent.properties[kBCOVPlaybackSessionEventKeyError] as? NSError {
            // Report any errors that may have occurred with playback.
            print("BaseViewController - Playback error: \(error.localizedDescription)")
        }

        // Ad events are emitted by the BCOVDAI plugin through lifecycle events.
        // The events are defined in BCOVDAIComponent.h.
        if kBCOVDAILifecycleEventAdsLoaderLoaded == lifecycleEvent.eventType,
           let adsManager = lifecycleEvent.properties[kBCOVDAILifecycleEventPropertyKeyAdsManager] as? IMAAdsManager {
            // Lower the volume of ads by half.
            adsManager.volume = adsManager.volume / 2.0
            print("BaseViewController - IMAAdsManager.volume set to \(String(format: "%0.1f", adsManager.volume))")

        } else if kBCOVDAILifecycleEventAdsManagerDidReceiveAdEvent == lifecycleEvent.eventType,
                  let adEvent = lifecycleEvent.properties[kBCOVDAILifecycleEventPropertyKeyAdEvent] as? IMAAdEvent {
            switch adEvent.type {
                case .STARTED:
                    adIsPlaying = true
                case .COMPLETE:
                    adIsPlaying = false
                default:
                    break
            }
        }
    }
}


// MARK: - BCOVPlaybackControllerAdsDelegate

extension BaseViewController: BCOVPlaybackControllerAdsDelegate {

    func playbackController(_ controller: BCOVPlaybackController,
                            playbackSession session: BCOVPlaybackSession,
                            didEnterAdSequence adSequence: BCOVAdSequence) {
        print("ViewController - Entering ad sequence")
    }

    func playbackController(_ controller: BCOVPlaybackController,
                            playbackSession session: BCOVPlaybackSession,
                            didExitAdSequence adSequence: BCOVAdSequence) {
        print("ViewController - Exiting ad sequence")
    }

    func playbackController(_ controller: BCOVPlaybackController,
                            playbackSession session: BCOVPlaybackSession,
                            didEnterAd ad: BCOVAd) {
        print("ViewController - Entering ad")
    }

    func playbackController(_ controller: BCOVPlaybackController,
                            playbackSession session: BCOVPlaybackSession,
                            didExitAd ad: BCOVAd) {
        print("ViewController - Exiting ad")
    }
}


// MARK: - BCOVPUIPlayerViewDelegate

extension BaseViewController: BCOVPUIPlayerViewDelegate {

    func playerView(_ playerView: BCOVPUIPlayerView!,
                    willTransitionTo screenMode: BCOVPUIScreenMode) {
        statusBarHidden = screenMode == .full
    }
}


// MARK: - BCOVDAIPlaybackSessionDelegate

extension BaseViewController: BCOVDAIPlaybackSessionDelegate {

    func willCallIMAAdsLoaderRequestAds(with adsRequest: IMAStreamRequest) {
        // for demo purposes, modify the adTagParameters
        // https://support.google.com/admanager/answer/7320899?hl=en

        adsRequest.adTagParameters = [ "tfcd": "0" ]
    }
}


// MARK: - IMALinkOpenerDelegate

extension BaseViewController: IMALinkOpenerDelegate {

    func linkOpenerDidClose(inAppLink linkOpener: NSObject) {
        // Called when the in-app browser has closed.
        guard let playbackController else { return }
        playbackController.resumeAd()
    }
}
